### PR TITLE
Fix infinite loop in dead-branch-elimination

### PR DESCRIPTION
**Pull request by s-perron**
* Create structed cfg analysis.

There are lots of optimization that have to traverse the CFG in a
structured order just because it wants to know which constructs a
basic block in contained in.  This adds extra complexity to these
optimizations, for causes too much refactoring of older optimizations.

To help with this problem, I have written an analysis that can give this
information.

* Identify branches breaking from loops.

Dead branch elimination does a search for a conditional branch to the
end of the current selection construct.  This search assumes that the
only way to leave the construct is through the merge node.  But that is
not true.  The code can jump to the merge node of a loop that contains
the construct.

The search needs to take this into consideration.

### CODE REVIEW

**Comment by s-perron** [comment]
Still have to add unit tests.

**Comment by dnovillo** [comment]
So, you had added tests for the analysis but I missed them because github's UI hides large diffs.  Sorry!

File: source/opt/struct_cfg_analysis.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  // Returns the id of the merge block of the innermost loop construct
+  // that contains |bb_id|.
+  uint32_t LoopMergeBlock(uint32_t bb_id);
+
+ private:
+  struct ConstructInfo {
+    uint32_t containing_construct;
+    uint32_t containing_loop;
```
[the remark this answers is not in the document]
Comment (s-perron) [reply source/opt/struct_cfg_analysis.h]: Not when the basic block is inside an if-statement that is inside a loop.  In this case, containing_construct will be the header for the selection merge, and the containing loop will be the header for the loop merge.

[the remark this answers is not in the document]
Comment (s-perron) [reply source/opt/struct_cfg_analysis.h]: Added comments to describe |ConstructInfo|.  That should help clarify.

**Pull request merged**